usm: kafka: Support fetch v12

vitkyrka opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

Add support for v12 of the Fetch API.  The main change is the use of
varints and addition of tagged values.  While that change itself is
fairly straightforward, the need to parse varints caused the programs to
exceed the instruction size supported by older kernels (4096
instructions), so the programs had to be split into separate programs
for pre-v12 and v12, as well as to do the parsing in two steps in
separate programs: a first step to parse the locations of the record
batches arrays, and a second step to parse the list of found locations.

The split between pre-v12 and v12 is simple and handled by informing the 
compiler about the bounds of the versions, and then depending on it to
eliminate the unused branches.  The split of the parsing into two steps
requires code to handle the transition between the two steps correctly.

Also mainly due to code size restrictions on older kernels (even after
the above optimizations), the following limitations are currently
present:

 - A maximum of four bytes of varints are supported
 - The tagged values fields are always assumed to be empty


<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

https://datadoghq.atlassian.net/browse/USMON-829

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

#### Maps change

`kafka_response` map increased from 6.92 MiB to 7.15 MiB (with 30k entries).

#### Load test results

Base version is main (`c4dd4ac`). Test type is `kafka.plain`.

![system-probe max k8s CPU Usage  (3)](https://github.com/DataDog/datadog-agent/assets/465294/23421505-59d3-45f9-875a-ebf7a7a761dd)
![system-probe RSS (avg) (3)](https://github.com/DataDog/datadog-agent/assets/465294/57d84311-1974-433d-bf22-cb37d36cc3a9)
![process agent working set (avg)](https://github.com/DataDog/datadog-agent/assets/465294/08b2ed05-fc68-4ac9-90a6-5ffa605142d4)

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

vitkyrka committed:
usm: kafka: Use unsigned loop counter

This allows some compiler optimzations and avoids a "looks like the BPF
stack limit of 512 bytes is exceeded" error when more code is added to
the classification functions.

vitkyrka committed:
usm: kafka: test: Make unused fields non-zero

Set some unparsed fields in the packets to non-zero values for easier
debugging when packet parsing goes wrong and reads from wrong offsets.

vitkyrka committed:
usm: kafka: Support Fetch v12

Add support for v12 of the Fetch API.  The main change is the use of
varints and addition of tagged values.  While that change itself is
fairly straightforward, the need to parse varints caused the programs to
exceed the instruction size supported by older kernels (4096
instructions), so the programs had to be split into separate programs
for pre-v12 and v12, as well as to do the parsing in two steps in
separate programs: a first step to parse the locations of the record
batches arrays, and a second step to parse the list of found locations.

The split between pre-v12 and v12 is simple and handled by informing the
compiler about the bounds of the versions, and then depending on it to
eliminate the unused branches.  The split of the parsing into two steps
requires code to handle the transition between the two steps correctly.

Also mainly due to code size restrictions on older kernels (even after
the above optimizations), the following limitations are currently
present:

 - A maximum of four bytes of varints are supported
 - The tagged values fields are always assumed to be empty

vitkyrka committed:
usm: kafka: test: Pass FetchRequest to buildMessages

To allow the request to be customized per test.

vitkyrka committed:
usm: kafka: test: Pass in API version

Pass in the API version as a parameter to the raw tests, so that
different API versions can be tested.

vitkyrka committed:
usm: kafka: test: Test different versions in raw test

Test some versions which have specific conditions.

vitkyrka committed:
usm: kafka: test: Allow topic to be customized

vitkyrka committed:
usm: kafka: test: Add test for large topic name

vitkyrka committed:
usm: kafka: test: Add test for many topics

vitkyrka committed:
usm: kafka: test: Add raw test for many partitions

vitkyrka committed:
usm: kafka: test: Handle flexible header

vitkyrka committed:
usm: kafka: Pass in version number to parsing tests

DanielLavie commented on pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:
```diff
@@ -255,14 +255,187 @@ static __always_inline enum parse_result read_with_remainder(kafka_response_cont
                                                                             pktbuf_t pkt, u32 offset,
-                                                                            u32 data_end)
+                                                                            u32 data_end,
+                                                                            enum parser_level level,
```
I don't see `level` being used here

DanielLavie commented on pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:
```diff
@@ -366,21 +571,105 @@ static __always_inline enum parse_result kafka_continue_parse_response_loop(kafk
+    return RET_LOOP_END;
+}
+
+static __always_inline enum parse_result kafka_continue_parse_response_record_batches_loop(kafka_info_t *kafka,
```
I don't see `level` being used here

DanielLavie commented on pkg/network/ebpf/c/protocols/kafka/kafka-parsing.h:
```diff
@@ -526,15 +839,150 @@ static __always_inline enum parse_result kafka_continue_parse_response_loop(kafk
+static __always_inline void kafka_call_response_parser(void *ctx, conn_tuple_t *tup, pktbuf_t pkt, kafka_response_state state, u32 api_version)
+{
+    enum parser_level level = parser_state_to_level(state);
+    u32 index;
```
```suggestion
    u32 index = 0;
```

DanielLavie replied:
I see, can you please add a comment in the code clarifying this point?

vitkyrka replied:
Done.

DanielLavie commented on pkg/network/ebpf/c/protocols/kafka/types.h:
```diff
@@ -69,13 +81,24 @@ typedef struct kafka_response_context_t {
+    // KAFKA_FETCH_RESPONSE_RECORD_BATCHES_ARRAY_END.
+    __u32 record_batches_arrays_idx;
+    __u32 record_batches_arrays_count;
+    // kafka_fetch_response_record_batches_array saved_record_batch_array;
```
Is this a leftover?

DanielLavie commented on pkg/network/protocols/kafka/types.go:
```diff
@@ -21,6 +21,7 @@ type ConnTuple C.conn_tuple_t
 
 type EbpfTx C.kafka_event_t
 
+type KafkaFetchResponseRecordBatchesArray C.kafka_fetch_response_record_batches_array_t
```
I don't see a usage of this type

DanielLavie commented on pkg/network/usm/kafka_monitor_test.go:
```diff
@@ -168,6 +190,9 @@ func (s *KafkaProtocolParsingSuite) testKafkaProtocolParsing(t *testing.T, tls b
 		return getDefaultTestConfiguration(tls)
 	}
 
+	tmp, _ := version.LookupMaxKeyVersion(kafka.FetchAPIKey)
```
Could you incorporate handling for the scenario where the function returns false?

vitkyrka committed:
usm: kafka: test: Check for errors from LookupMaxKeyVersion

vitkyrka committed:
usm: kafka: Remove leftover comment

vitkyrka committed:
usm: kafka: remove unused cgo type

vitkyrka committed:
usm: kafka: Remove passing of level

vitkyrka committed:
usm: kafka: Add comment to parse_varint_u16()

vitkyrka committed:
usm: kafka: Fix u16 varint check

The second byte's MSB should only be checked if the first byte has the
MSB set, since otherwise the second byte is unrelated to the varint.

vitkyrka committed:
usm: kafka: Rename topic_name_size_tmp to topic_name_size_raw

vitkyrka committed:
usm: kafka: Handle splits in header

We were previously not handling this mainly due to code size
restrictions, but with the program split this can be supported.

vitkyrka committed:
usm: kafka: Make read_with_remainder generic

Remove the copy/pasted code and make read_with_remainder() generic with
the use of a function pointer.  The function pointer will be folded at
compile time by the compiler.  The instruction count is unchanged before
and after this patch.

vitkyrka committed:
usm: kafka: Remove unused nonrestartable functions

vitkyrka committed:
usm: kafka: Remove _restartable suffix

Now that we only have restartable readers, remove the redundant suffix.

vitkyrka committed:
usm: kafka: Validate zero tagged fields when possible

vitkyrka committed:
usm: kafka: Remove __read_varint indirection

vitkyrka committed:
usm: kafka: Allow varint bytes to be customized per location

To reduce code size by limiting iterations wherever it makes sense.

vitkyrka committed:
usm: kafka: Customize varint bytes per location

vitkyrka committed:
usm: kafka: Clarify N-1 operation in varints

vitkyrka committed:
usm: kafka: Rename tmp2 variable

vitkyrka committed:
usm: kafka: Add comment about flexible

vitkyrka committed:
usm: kafka: Skip varint number of topics

vitkyrka committed:
usm: kafka: Check zero tagged fields in request

vitkyrka committed:
usm: kafka: Use 64 bits for varint results

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into kafka-fetch-v12

DanielLavie reviewed (approved):
Left a single nit comment

vitkyrka committed:
usm: kafka: Add comment about uninitialized variable

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into kafka-fetch-v12

 Conflicts:
	pkg/network/usm/kafka_monitor_test.go

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into kafka-fetch-v12

vitkyrka committed:
usm: kafka: test: Fix teardown

The telemetry test assigns to the client in the context multiple times,
so some of the clients are never torn down, leading to issues with tests
run later (such as for example the still-running clients attempting to
connect to the raw server while the raw tests are executed afterward).

vitkyrka committed:
Merge remote-tracking branch 'origin/main' into kafka-fetch-v12

vitkyrka committed:
Merge branch 'kafka-test-fix-teardown' into kafka-fetch-v12

vitkyrka commented:
/merge
